staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -468,26 +474,37 @@ func (qs *queueSet) timeoutOldRequestsAndRejectOrEnqueueLocked(ctx context.Conte
 	return req
 }
 
-// chooseQueueIndexLocked uses shuffle sharding to select a queue index
+// shuffleShardLocked uses shuffle sharding to select a queue index
 // using the given hashValue and the shuffle sharding parameters of the queueSet.
-func (qs *queueSet) chooseQueueIndexLocked(hashValue uint64, descr1, descr2 interface{}) int {
+func (qs *queueSet) shuffleShardLocked(hashValue uint64, descr1, descr2 interface{}) int {
+	var backHand [8]int
+	// Deal into a data structure, so that the order of visit below is not necessarily the order of the deal.
+	// This removes bias in the case of flows with overlapping hands.
+	hand := qs.dealer.DealIntoHand(hashValue, backHand[:])
+	handSize := len(hand)
+	offset := qs.enqueues % handSize
+	qs.enqueues++
 	bestQueueIdx := -1
 	bestQueueSeatsSum := int(math.MaxInt32)
-	// the dealer uses the current desired number of queues, which is no larger than the number in `qs.queues`.
-	qs.dealer.Deal(hashValue, func(queueIdx int) {
+	for i := 0; i < handSize; i++ {
+		queueIdx := hand[(offset+i)%handSize]
+		queue := qs.queues[queueIdx]
+		waitingSeats := queue.requests.SeatsSum()
 		// TODO: Consider taking into account `additional latency` of requests
 		// in addition to their seats.
 		// Ideally, this should be based on projected completion time in the
 		// virtual world of the youngest request in the queue.
-		queue := qs.queues[queueIdx]
-		waitingSeats := queue.requests.SeatsSum()
-		thisSeatsSum := waitingSeats // + queue.seatsInUse
-		klog.V(7).Infof("QS(%s): For request %#+v %#+v considering queue %d of %d seats waiting and %d executing", qs.qCfg.Name, descr1, descr2, queueIdx, waitingSeats, queue.seatsInUse)
+		thisSeatsSum := waitingSeats + queue.seatsInUse
```

Comment:
@tkashem - I think you commented out seatsInUse in the past, but I can't remember the exact reason. Can you remind this?

Reply:
It was giving worse results.
With the other changes, it now gives good enough results.

Reply:
this line was added by Mike's PR https://github.com/kubernetes/kubernetes/pull/104345/files#diff-9411bfa852eec5372d491bd393f19ef57cab8d5cd54af3a842754dcc327f68a7R482-R484. 
```
queue := qs.queues[queueIdx]
waitingSeats := queue.requests.SeatsSum()
thisSeatsSum := waitingSeats // + queue.seatsInUse
```

I believe un-commenting `queue.seatsInUse` makes sense here.

Reply:
SGTM